# Combining DataFrames With Pandas - Lab

## Introduction

In this lab, you'll gain practice combining DataFrames through concatenation.  You'll also practice executing various types of joins to selectively combine the information stored in the tables!

## Objectives

In this lab you will:

- Use concatenation to combine DataFrames  
- Determine which type of join is preferred for two tables of data and a task  
- Use different types of joins to merge dataframes


## Concatenating DataFrames

Run the cell below to create some sample DataFrames for us to work with.  

In [1]:
import pandas as pd
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                    index=[0, 1, 2, 3])


df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                    index=[4, 5, 6, 7])

df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'], 
                    'D': ['D8', 'D9', 'D10', 'D11']},
                    index=[8, 9, 10, 11])

In [3]:
df1.head()

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [4]:
df2.head()

A   B   C   D
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [5]:
df3.head()

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

Now that you have multiple DataFrames to work with, you can execute a concatenation to join them together.  

In the cell below, concatenate the 3 DataFrames together using the appropriate function.   

In [6]:
combined_df = pd.concat([df1, df2, df3], axis = 1)
combined_df
# refer https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html 
# if you do shift tab inside the parenthesis you will see details & 
# apart from objects everything else has an equalto sign that means objects is the required  parameter to be put in & 
# in this case i need to return has to be a list 

A    B    C    D    A    B    C    D    A    B    C    D
0    A0   B0   C0   D0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   B4   C4   D4  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   A5   B5   C5   D5  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   A6   B6   C6   D6  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11

In [7]:
combined_df = pd.concat([df1, df2, df3], axis = 0) # when you change the axis to zero 
# axis{0/’index’, 1/’columns’}, default 0
# The axis to concatenate along.
combined_df

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

**_EXPECTED OUTPUT:_**

<img src="images/er1.png">

## Setting join conditions with concatenation

You can also specify if the concatenation is an **_Outer Join_** or an **_Inner Join_**.  Next, you'll execute an inner join. Before you do, you need to create another table that contains some overlapping index values with a DataFrame that already exists. 

Run the cell below to create the new DataFrame.

In [8]:
df4 = pd.DataFrame({'B': ['B2', 'B3', 'B6', 'B7'],
                    'D': ['D2', 'D3', 'D6', 'D7'],
                    'F': ['F2', 'F3', 'F6', 'F7']},
                    index=[2, 3, 6, 7]) # this is making a fourth data frame

In [10]:
df4

B   D   F
2  B2  D2  F2
3  B3  D3  F3
6  B6  D6  F6
7  B7  D7  F7

Now, in the cell below, use the `pd.concat()` function to join DataFrames 1 and 4.  However, this time, specify that the `join` is `'inner'`, and `axis=1`. 

In [17]:
df1_and_4 = pd.concat([df1,df4], join='inner', axis=1) 
# here we are combining df1 & df4 and we are specifying inner join and axis = 1
df1_and_4

A   B   C   D   B   D   F
2  A2  B2  C2  D2  B2  D2  F2
3  A3  B3  C3  D3  B3  D3  F3

In [18]:
df1_and_4 = pd.concat([df1,df4], join='outer', axis=1) 
# here we are combining df1 & df4 and we are specifying outer join and axis = 1


# df1_and_4 = pd.concat([df1,df4], join='left') 
# df1_and_4
#the join="left" will show this ValueError: Only can inner (intersect) or outer (union) join the other axis




df1_and_4

A    B    C    D    B    D    F
0   A0   B0   C0   D0  NaN  NaN  NaN
1   A1   B1   C1   D1  NaN  NaN  NaN
2   A2   B2   C2   D2   B2   D2   F2
3   A3   B3   C3   D3   B3   D3   F3
6  NaN  NaN  NaN  NaN   B6   D6   F6
7  NaN  NaN  NaN  NaN   B7   D7   F7

In [16]:
df1n4 = df1.append(df4, sort=False)
df1n4

A   B    C   D    F
0   A0  B0   C0  D0  NaN
1   A1  B1   C1  D1  NaN
2   A2  B2   C2  D2  NaN
3   A3  B3   C3  D3  NaN
2  NaN  B2  NaN  D2   F2
3  NaN  B3  NaN  D3   F3
6  NaN  B6  NaN  D6   F6
7  NaN  B7  NaN  D7   F7

In [14]:
df1_and_4 = pd.concat([df1,df4], join='left') 
# here we are combining df1 & df4 and we are specifying left join and axis = 1
df1_and_4

ValueError: Only can inner (intersect) or outer (union) join the other axis

**_EXPECTED OUTPUT:_**

<img src='images/er2.png'>

You'll notice that in this case, the results contain only the rows with indexes that exist in both tables -- rows 2 and 3.  The resulting table contains the values for each column in both tables for the rows.  

Note that there are many, many ways that you can make full use of the `pd.concat()` function in pandas to join DataFrames together -- these are just a few of the most common examples pulled from the pandas documentation on the subject. For a full view of all the ways you can use `pd.concat()`, see the [pandas documentation](http://pandas.pydata.org/pandas-docs/stable/merging.html)!

## Load data
Now, it's time to move on to working with the Hearthstone cards database.  This database contains information on cards from the popular game, [Hearthstone](https://playhearthstone.com/en-us/)! For full information on the dataset, see the [Kaggle page](https://www.kaggle.com/jeradrose/hearthstone-cards) for this dataset. 

This database consists of the following tables:

* _cards_
* *dust_costs*
* _entourages_
* _mechanics_
* *play_requirements*

Many of rows in each table -- but not all -- correspond to the same cards. As such, each table contains a column called `card_id` which acts as a **_Primary Key_** for each table.  You'll make use of these keys to **_join_** the different tables together into a single DataFrame. You'll also experiment with different types of joins to help us decide exactly what information you wish to combine.  

Simply run the cell below to import the tables from the database as DataFrames.

In [40]:
cards_df = pd.read_csv('cards.csv')
dust_df = pd.read_csv('dust.csv')
entourages_df = pd.read_csv('entourages.csv')
mechanics_df = pd.read_csv('mechanics.csv')
play_requirements_df = pd.read_csv('play_requirements.csv')

Great.  Now, let's set the correct column, `card_id`, as the index column for each of these tables, and then display each to ensure that everything is as expected.  

For each of the DataFrames you created in the cell above, call the `.set_index()` method and pass in `card_id`.  Also set `inplace=True`.  Then, display the `.head()` of each respective DataFrame to ensure everything worked.  

**_NOTE:_** Since you are performing this operation in place, running any cell a second time will result in pandas throwing an error.  If you need to run something a second time, restart the kernel using the jupyter notebook menu at the top of the page.  

In [21]:
cards_df.set_index('card_id')

player_class         type                                    name  \
card_id                                                                        
KARA_00_07         MAGE        SPELL                           Astral Portal   
NEW1_008a         DRUID        SPELL                       Ancient Teachings   
BRM_010t2         DRUID       MINION                      Druid of the Flame   
AT_132          NEUTRAL       MINION                      Justicar Trueheart   
OG_141          NEUTRAL       MINION                       Faceless Behemoth   
...                 ...          ...                                     ...   
AT_066e         WARRIOR  ENCHANTMENT                     Forges of Orgrimmar   
CS2_031            MAGE        SPELL                               Ice Lance   
OG_195e         NEUTRAL  ENCHANTMENT                                Enormous   
XXX_111e        NEUTRAL  ENCHANTMENT  All Charge, All Windfury, All The Time   
NAX3_01         NEUTRAL         HERO                                 Maexxna   

                set                                               text  cost  \
card_id                                                                        
KARA_00_07     KARA           Summon a random <b>Legendary</b> minion.   1.0   
NEW1_008a   EXPERT1                                       Draw a card.   0.0   
BRM_010t2       BRM                                                NaN   3.0   
AT_132          TGT  <b>Battlecry:</b> Replace your starting Hero P...   6.0   
OG_141           OG                                                NaN  10.0   
...             ...                                                ...   ...   
AT_066e         TGT                                  Increased Attack.   NaN   
CS2_031     EXPERT1  <b>Freeze</b> a character. If it was already <...   1.0   
OG_195e          OG                                             +2/+2.   NaN   
XXX_111e      CHEAT  Your minions always have <b>Charge</b> and <b>...   NaN   
NAX3_01        NAXX                                                NaN   NaN   

            attack  health     rarity  collectible  \
card_id                                              
KARA_00_07     NaN     NaN        NaN          NaN   
NEW1_008a      NaN     NaN        NaN          NaN   
BRM_010t2      2.0     5.0     COMMON          NaN   
AT_132         6.0     3.0  LEGENDARY          1.0   
OG_141        10.0    10.0     COMMON          1.0   
...            ...     ...        ...          ...   
AT_066e        NaN     NaN        NaN          NaN   
CS2_031        NaN     NaN     COMMON          1.0   
OG_195e        NaN     NaN        NaN          NaN   
XXX_111e       NaN     NaN        NaN          NaN   
NAX3_01        NaN    30.0        NaN          NaN   

                                                       flavor   race  \
card_id                                                                
KARA_00_07                                                NaN    NaN   
NEW1_008a                                                 NaN    NaN   
BRM_010t2                                                 NaN  BEAST   
AT_132      It's like putting racing stripes and a giant s...    NaN   
OG_141      Rejected names: Forty-Foot Faceless, Big ol' N...    NaN   
...                                                       ...    ...   
AT_066e                                                   NaN    NaN   
CS2_031     The trick is not to break the lance.  Otherwis...    NaN   
OG_195e                                                   NaN    NaN   
XXX_111e                                                  NaN    NaN   
NAX3_01                                                   NaN    NaN   

           how_to_earn how_to_earn_golden targeting_arrow_text faction  \
card_id                                                                  
KARA_00_07         NaN                NaN                  NaN     NaN   
NEW1_008a          NaN                NaN                  NaN     

In [23]:
cards_df.head(2)

card_id player_class   type               name      set  \
0  KARA_00_07         MAGE  SPELL      Astral Portal     KARA   
1   NEW1_008a        DRUID  SPELL  Ancient Teachings  EXPERT1   

                                       text  cost  attack  health rarity  \
0  Summon a random <b>Legendary</b> minion.   1.0     NaN     NaN    NaN   
1                              Draw a card.   0.0     NaN     NaN    NaN   

   collectible flavor race how_to_earn how_to_earn_golden  \
0          NaN    NaN  NaN         NaN                NaN   
1          NaN    NaN  NaN         NaN                NaN   

  targeting_arrow_text faction  durability  
0                  NaN     NaN         NaN  
1                  NaN     NaN         NaN

In [41]:
cards_df.set_index('card_id', inplace=True)
cards_df.head(2)

player_class   type               name      set  \
card_id                                                      
KARA_00_07         MAGE  SPELL      Astral Portal     KARA   
NEW1_008a         DRUID  SPELL  Ancient Teachings  EXPERT1   

                                                text  cost  attack  health  \
card_id                                                                      
KARA_00_07  Summon a random <b>Legendary</b> minion.   1.0     NaN     NaN   
NEW1_008a                               Draw a card.   0.0     NaN     NaN   

           rarity  collectible flavor race how_to_earn how_to_earn_golden  \
card_id                                                                     
KARA_00_07    NaN          NaN    NaN  NaN         NaN                NaN   
NEW1_008a     NaN          NaN    NaN  NaN         NaN                NaN   

           targeting_arrow_text faction  durability  
card_id                                              
KARA_00_07                  NaN     NaN         NaN  
NEW1_008a                   NaN     NaN         NaN

In [42]:
dust_df.set_index('card_id', inplace=True)
dust_df.head(2)

action  cost
card_id                         
BRM_010t2  CRAFTING_NORMAL    40
BRM_010t2  CRAFTING_GOLDEN   400

In [43]:
entourages_df.set_index('card_id', inplace=True)
entourages_df.head(2)

entourage_card_id
card_id                     
KAR_A10_22        KAR_A10_09
KAR_A10_22        KAR_A10_02

In [44]:
mechanics_df.set_index('card_id', inplace=True)
mechanics_df.head(2)

mechanic
card_id                      
AT_132              BATTLECRY
GVG_011a  TAG_ONE_TURN_EFFECT

In [45]:
play_requirements_df.set_index('card_id', inplace=True)
play_requirements_df.head(2)

play_requirement  value
card_id                                
KARA_00_07  REQ_NUM_MINION_SLOTS      1
PRO_001a    REQ_NUM_MINION_SLOTS      1

## Executing Joins

Now that you have the tables loaded correctly, we're going to execute some joins. There are four different kinds of joins, which can best be visualized with Venn diagrams:

<img src='images/Image_198_joins.png'>

In these diagrams, each circle represents a DataFrame or SQL Table. The left table is the table you are working with, and the right table is the table you want to join to the table you are working with. You'll start by executing the most common type of join, an **_Inner Join_**.

In the cell below, join `cards_df` with `mechanics_df` using the built-in `.join()` method on the `cards_df` object. 

Pass in the following parameters:
* the table you want to join with, `mechanics_df`
* The `how` parameter set to the type of join you want, `'inner'`

In [46]:
cards_with_mechanics_df = cards_df.join(mechanics_df, how='inner')
cards_with_mechanics_df.head(2)

player_class    type    name  set  \
card_id                                     
AT_002          MAGE   SPELL  Effigy  TGT   
AT_005t      NEUTRAL  MINION    Boar  TGT   

                                                      text  cost  attack  \
card_id                                                                    
AT_002   <b>Secret:</b> When a friendly minion dies, su...   3.0     NaN   
AT_005t                                      <b>Charge</b>   3.0     4.0   

         health rarity  collectible              flavor   race how_to_earn  \
card_id                                                                      
AT_002      NaN   RARE          1.0  Burning man, brah.    NaN         NaN   
AT_005t     2.0    NaN          NaN                 NaN  BEAST         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability mechanic  
card_id                                                                       
AT_002                 NaN                  NaN     NaN         NaN   SECRET  
AT_005t                NaN                  NaN     NaN         NaN   CHARGE

In [47]:
cards_with_mechanics_df = cards_df.join(mechanics_df, how='inner')
cards_with_mechanics_df.head(2)

player_class    type    name  set  \
card_id                                     
AT_002          MAGE   SPELL  Effigy  TGT   
AT_005t      NEUTRAL  MINION    Boar  TGT   

                                                      text  cost  attack  \
card_id                                                                    
AT_002   <b>Secret:</b> When a friendly minion dies, su...   3.0     NaN   
AT_005t                                      <b>Charge</b>   3.0     4.0   

         health rarity  collectible              flavor   race how_to_earn  \
card_id                                                                      
AT_002      NaN   RARE          1.0  Burning man, brah.    NaN         NaN   
AT_005t     2.0    NaN          NaN                 NaN  BEAST         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability mechanic  
card_id                                                                       
AT_002                 NaN                  NaN     NaN         NaN   SECRET  
AT_005t                NaN                  NaN     NaN         NaN   CHARGE

In [48]:
cards_with_mechanics_df.shape

(1079, 18)

In [51]:
cards_df.shape

(2819, 17)

In [52]:
mechanics_df.shape

(1079, 1)

In [53]:
mechanics_df.head(2)

mechanic
card_id                      
AT_132              BATTLECRY
GVG_011a  TAG_ONE_TURN_EFFECT

Examine the output from the cell above and compare it to the original output of both the `cards_df` and `mechanics_df` DataFrame heads you displayed earlier.  Notice how this now combines the columns from both?

**_Question_**

If you inspect the original `cards_df` DataFrame, you'll notice that it contains  2,819 records.  The result of our inner join, `cards_with_mechanics_df`, contains only 1079 rows.  Why?

Write your answer below this line:
________________________________________________________________________________

## Other Types of Joins

By default, the `.join()` method performs a left join if no parameter is passed in for `how=`.  In the cell below, perform a **_Left Join_** of `cards_with_mechanics_df` and `play_requirements_df`, with `cards_with_mechanics_df` as the left table.  

Then, display `left_join_df` to inspect our results. 

In [54]:
left_join_df = cards_with_mechanics_df.join(play_requirements_df, how='left')
left_join_df

player_class    type                    name      set  \
card_id                                                         
AT_002          MAGE   SPELL                  Effigy      TGT   
AT_005t      NEUTRAL  MINION                    Boar      TGT   
AT_006          MAGE  MINION        Dalaran Aspirant      TGT   
AT_007          MAGE  MINION            Spellslinger      TGT   
AT_009          MAGE  MINION                  Rhonin      TGT   
...              ...     ...                     ...      ...   
XXX_105      NEUTRAL   SPELL        Add 8 to Health.    CHEAT   
XXX_107      NEUTRAL   SPELL         Set Health to 1    CHEAT   
XXX_110      NEUTRAL  MINION  Yogg-Saron Test (Auto)    CHEAT   
hexfrog      NEUTRAL  MINION                    Frog     CORE   
tt_010          MAGE   SPELL             Spellbender  EXPERT1   

                                                      text  cost  attack  \
card_id                                                                    
AT_002   <b>Secret:</b> When a friendly minion dies, su...   3.0     NaN   
AT_005t                                      <b>Charge</b>   3.0     4.0   
AT_006        <b>Inspire:</b> Gain <b>Spell Damage +1</b>.   4.0     3.0   
AT_007   <b>Battlecry:</b> Add a random spell to each p...   3.0     3.0   
AT_009   <b>Deathrattle:</b> Add 3 copies of Arcane Mis...   8.0     7.0   
...                                                    ...   ...     ...   
XXX_105  Adds 8 health to a damaged character. Does NOT...   0.0     NaN   
XXX_107  Set a character's health to 1, and remove all ...   0.0     NaN   
XXX_110  <b>Battlecry:</b> Cast 30 random spells <i>(ta...   0.0     7.0   
hexfrog                                       <b>Taunt</b>   0.0     0.0   
tt_010   <b>Secret:</b> When an enemy casts a spell on ...   3.0     NaN   

         health     rarity  collectible  \
card_id                                   
AT_002      NaN       RARE          1.0   
AT_005t     2.0        NaN          NaN   
AT_006      5.0     COMMON          1.0   
AT_007      4.0     COMMON          1.0   
AT_009      7.0  LEGENDARY          1.0   
...         ...        ...          ...   
XXX_105     NaN     COMMON          NaN   
XXX_107     NaN     COMMON          NaN   
XXX_110     5.0  LEGENDARY          NaN   
hexfrog     1.0     COMMON          NaN   
tt_010      NaN       EPIC          1.0   

                                                    flavor   race how_to_earn  \
card_id                                                                         
AT_002                                  Burning man, brah.    NaN         NaN   
AT_005t                                                NaN  BEAST         NaN   
AT_006    Is he aspiring or inspiring?  Make up your mind!    NaN         NaN   
AT_007   Does he sling spells, or do his spells linger ...    NaN         NaN   
AT_009                              A masterless shamurai.    NaN         NaN   
...                                                    ...    ...         ...   
XXX_105                                                NaN    NaN         NaN   
XXX_107                                                NaN    NaN         NaN   
XXX_110                                                NaN    NaN         NaN   
hexfrog                                                NaN  BEAST         NaN   
tt_010   While it's fun to intercept enemy lightning bo...    NaN         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability  \
card_id                                                               
AT_002                 NaN                  NaN     NaN         NaN   
AT_005t                NaN                  NaN     NaN         NaN   
AT_006                 NaN                  NaN     NaN         NaN   
AT_007                 NaN                  NaN     NaN         NaN   
AT_009                 NaN                  NaN     NaN         NaN   
...                    ...                  ...     ...       

Note that the results of this sort of join are dependent upon the position of each table--if you were to make `cards_with_mechanics_df` the right table and `play_requirements_df` the left table and then perform a **_Right Join_**, our results would be the same. 

**_Question:_**

Describe what was included from each table in this join.

Write your answer below this line:
________________________________________________________________________________

#### Outer Joins

In the cell below, perform an outer join between `cards_df` and `dust_df`. Since these tables contain columns with the same name, we'll need to specify a suffix for at least one of them, so that the column can be renamed to avoid a naming collision. 

During this join, set the `rsuffix` parameter to `_dust`

In [55]:
cards_df.columns

Index(['player_class', 'type', 'name', 'set', 'text', 'cost', 'attack',
       'health', 'rarity', 'collectible', 'flavor', 'race', 'how_to_earn',
       'how_to_earn_golden', 'targeting_arrow_text', 'faction', 'durability'],
      dtype='object')

In [56]:
dust_df.columns

Index(['action', 'cost'], dtype='object')

In [59]:
outer_join_df = cards_df.join(dust_df, how = 'outer', rsuffix='_dust')
outer_join_df.head(2)

player_class   type         name  set                         text  \
card_id                                                                      
AT_001          MAGE  SPELL  Flame Lance  TGT  Deal $8 damage to a minion.   
AT_001          MAGE  SPELL  Flame Lance  TGT  Deal $8 damage to a minion.   

         cost  attack  health  rarity  collectible  \
card_id                                              
AT_001    5.0     NaN     NaN  COMMON          1.0   
AT_001    5.0     NaN     NaN  COMMON          1.0   

                                                    flavor race how_to_earn  \
card_id                                                                       
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   
AT_001   It's on the rack next to ice lance, acid lance...  NaN         NaN   

        how_to_earn_golden targeting_arrow_text faction  durability  \
card_id                                                               
AT_001                 NaN                  NaN     NaN         NaN   
AT_001                 NaN                  NaN     NaN         NaN   

                  action  cost_dust  
card_id                              
AT_001   CRAFTING_NORMAL       40.0  
AT_001   CRAFTING_GOLDEN      400.0

Inspect the output above.  Note that the naming collision has been avoided by renaming the `cost` column from the right table to `cost_dust`.  

## Summary

In this lab, you learned how to:

* Concatenate multiple DataFrames together into a single DataFrame
* Understand and execute the various types of joins (inner, outer, left, and right joins)